### PR TITLE
Correctly fallback to HTTP api

### LINKED ISSUE #28: Get 403 error when query GitHub HTTP API on macOS
```
fetching master commit hash... 

error: https://api.github.com/repos/rust-lang/rust/commits/master: Client Error: 403 Forbidden
```

See this CI line: https://travis-ci.com/kennytm/rustup-toolchain-install-master/jobs/245113302#L444

### PR DESCRIPTION
**Pull request by lzutao**
Closes #28 .

### CODE REVIEW

**Comment by lzutao** [comment]
Hm, CI failed again. What's going on?

**Comment by lzutao** [comment]
CI is passed. Ping @kennytm .

File: src/main.rs [hunk 1]
```diff
@@ -313,12 +322,29 @@ fn fetch_master_commit_via_http(
...
+        StatusCode::FORBIDDEN => {
+            let rate_limit = response
+                .headers()
+                .get("X-RateLimit-Remaining")
+                .and_then(|r| r.to_str().ok())
+                .and_then(|r| r.parse::<u32>().ok())
+                .unwrap_or(0);
+            if rate_limit == 0 {
+                bail!("GitHub API rate limit exceeded")
+            }
```
Comment (lzutao) [inline src/main.rs]: I should check for else claise here

File: .travis.yml [hunk 2]
```diff
@@ -4,19 +4,25 @@ rust:
...
+  - osx
+  - windows
 
 cache:
   directories:
     - "$HOME/.cargo"
+    - "$TRAVIS_BUILD_DIR/target"
 before_cache:
-  - rm -rf "$HOME/.cargo/registry"
+  - travis_retry curl -Lf "https://github.com/lzutao/cargo-cache/releases/download/0.3.3/cargo-cache-0.3.3-${TRAVIS_OS_NAME}.tar.gz" | tar -xz --directory="$HOME/.cargo/bin"
```
Comment (kennytm) [inline .travis.yml]: ???

File: src/main.rs [hunk 3]
```diff
@@ -281,11 +288,13 @@ fn fetch_master_commit(client: &Client, github_token: Option<&str>) -> Result<St
...
     let res = fetch_master_commit_via_git()
         .context("unable to fetch master commit via git, falling back to HTTP");
-    if let Err(err) = res {
-        report_warn(&err);
+    match res {
+        Ok(commit) => Ok(commit),
+        Err(err) => {
+            report_warn(&err);
+            fetch_master_commit_via_http(client, github_token)
+        }
```
Comment (kennytm) [inline src/main.rs]: Let's use `or_else` here.

File: .travis.yml (hunk 2 above)
Comment (lzutao) [reply .travis.yml]: There is a PR https://github.com/matthiaskrgr/cargo-cache/pull/74, but it is not merged yet.
I use cargo-cache to cleanup ~/.cargo directory before caching it.

I will remove it and move it into another PR if you prefer.

Comment (lzutao) [reply .travis.yml]: Ping @kennytm. What is your thought on this?

Comment (kennytm) [reply .travis.yml]: Yes please remove the cache-related changes.

Comment (lzutao) [reply .travis.yml]: Done.

**Pull request merged**